### PR TITLE
Read VTK XML binary data [vtk-read-binary]

### PR DESCRIPTION
**Pull request by pazner**
Improve MFEM's VTK XML reader:

* Read appended raw data (small modifications to tinyxml2 required)
* Read appended base-64 encoded data (added base-64 decoding, may be useful for other features)
* Read inline base-64 encoded data
* Decompress zlib-compressed data
* Extract element attributes stored as "cell data" in VTK XML meshes

See [some sample meshes](https://github.com/mfem/mfem/files/6246505/quad.zip) for variations on these features.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#2150](https://github.com/mfem/mfem/pull/2150) | @pazner | @tzanio | @jandrej + @bslazarov | 04/01/21 | 04/13/21 | 04/16/21 | |

### CODE REVIEW

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by jandrej — APPROVED** [review]

**Comment by jandrej** [comment]
LGTM

**Comment by pazner** [comment]
@jandrej your LGTM failed because of the issue in mfem-uberenv that was fixed in #2151. Everything should pass now.

**Comment by bslazarov** [comment]
Tested and valgrinded - everything works as it should. Excellent work!

**Review by bslazarov — APPROVED** [review]

**Comment by tzanio** [comment]
LGTM

**Comment by tzanio** [comment]
@pazner, can you help with resolving the conflicts?

**Comment by pazner** [comment]
> @pazner, can you help with resolving the conflicts?

Done.

Also, sorry for the last-minute change, but it occurred to me that a unit test for this functionality would be very helpful. I added a small one to this PR in the latest commit. It involves adding some sample meshes to `tests/unit/data`. Hopefully the reviewers can take a quick glance to confirm this is ok, cc:  @jandrej @bslazarov

**Comment by bslazarov** [comment]
It seems that the tests failed otherwise I like the idea.

File: mesh/mesh.cpp [hunk 1]
```diff
@@ -3534,6 +3534,13 @@ void Mesh::Loader(std::istream &input, int generate_edges,
    getline(input, mesh_type);
    filter_dos(mesh_type);
 
+   // If first line is XML version, move on to second line
+   if (mesh_type == "<?xml version=\"1.0\"?>")
```
Comment (termi-official) [inline mesh/mesh.cpp:3538]: This is error prone as it just works if the XML is properly formatted, i.e. there is really a line break at the end of the opening tag.

**Review by termi-official — APPROVED** [review]
Really nice to have feature! I just have found something minor. Other than that LGTM.

File: mesh/mesh.cpp (hunk 1 above)
Comment (pazner) [reply mesh/mesh.cpp]: You're right — I think the handling of the `<?xml` declaration is improved now, and it should be able to handle the case where there is no following newline.

**Comment by pazner** [comment]
> It seems that the tests failed otherwise I like the idea.

Hopefully this is fixed now. There appears to really have been a bug that the tests caught, which I guess means that adding unit tests was the right decision :)

**Comment by tzanio** [comment]
@jandrej, @bslazarov and @termi-official can you confirm you are OK with the latest changes?

File: mesh/mesh_readers.cpp [hunk 2]
```diff
@@ -768,8 +768,10 @@ struct BufferReader : BufferReaderBase
       {
          for (int i=0; i<n; ++i)
          {
-            F *val = (F*)(buf + i*sizeof(F));
-            dest[i] = *val;
+            F val;
+            char *val_ptr = reinterpret_cast<char*>(&val);
+            std::copy(buf + i*sizeof(F), buf + (i+1)*sizeof(F), val_ptr);
+            dest[i] = val;
```
Comment (jandrej) [inline mesh/mesh_readers.cpp:774]: Isn't there a std library function to copy one entity of a certain type? This looks complicated for what it is doing.

Comment (pazner) [reply mesh/mesh_readers.cpp:774]: Not that I know of…

**Comment by termi-official** [comment]
@tzanio I will check on the weekend properly, but quickly going over the changes I think they should be fine.

**Comment by bslazarov** [comment]
I think that PR is good to go.

**Comment by tzanio** [comment]
Merged in `next` for testing...

**Pull request merged**